### PR TITLE
Introduce `pad_mode`

### PR DESCRIPTION
**Pull request by OverLordGoldDragon**
Allow user to choose between `reflect, symmetric, zero` padding.

While a fair default, `reflect` isn't all-purpose best; unless derivatives at boundaries are smooth, it may do more harm than good (I've seen in testing ridge extraction in `ssqueezepy`). Further, it is uniquely detrimental to JTFS, where lowpass captures opposite spin generated by `reflect` (confirmed with @lostanlen).

Current options are bottlenecked by TensorFlow; it can be configured per-backend. Most flexibly, a user can pass a `lambda`. For now I went with barebones that includes the important case of `zero` padding - most primitive, but also least priors, making it arguably general best. (P.S., answer to my 'quiz' at bottom [here](https://github.com/kymatio/kymatio/discussions/710) is boundary effects).

### CODE REVIEW

**Comment by lostanlen** [comment]
Cool. If you want i can write the unit tests for this?

**Comment by lostanlen** [comment]
Small nitpick but i would have liked `padding_mode` better than `padtype`.

The word `type` has a different connotation: `str`, `int`, `float`, etc.
Besides, NumPy has `mode`, Keras has `mode`, and TensorFlow has `mode`. 

So perhaps call it `mode` in internal `pad` functions and `padding_mode` in the constructor?

**Comment by changhongw** [comment]
> Small nitpick but i would have liked `padding_mode` better than `padtype`.
> 
> The word `type` has a different connotation: `str`, `int`, `float`, etc.
> Besides, NumPy has `mode`, Keras has `mode`, and TensorFlow has `mode`.
> 
> So perhaps call it `mode` in internal `pad` functions and `padding_mode` in the constructor?

Vote for `padding_mode`.

**Comment by changhongw** [comment]
> Allow user to choose between `reflect, symmetric, zero` padding.
> 
> While a fair default, `reflect` isn't all-purpose best; unless derivatives at boundaries are smooth, it may do more harm than good (I've seen in testing ridge extraction in `ssqueezepy`). Further, it is uniquely detrimental to JTFS, where lowpass captures opposite spin generated by `reflect` (confirmed with @lostanlen).
> 
> Current options are bottlenecked by TensorFlow; it can be configured per-backend. Most flexibly, a user can pass a `lambda`. For now I went with barebones that includes the important case of `zero` padding - most primitive, but also least priors, making it arguably general best. (P.S., answer to my 'quiz' at bottom [here](https://github.com/kymatio/kymatio/discussions/710) is boundary effects).
> 
> ### Todo
> * Docs (frontend): unsure where to "put" it, docstr layout is strange
> * Testing: unsure if I missed anything in some frontend

Nice idea in the long run!

**Comment by OverLordGoldDragon** [comment]
> unit tests

Sure

> `padtype`

Agreed; how about `padmode`, or `pad_mode`? `padding_mode` rather lengthy

**Comment by lostanlen** [comment]
`torch.Conv1D` has `padding_mode` so consistency with them would be nice


> `padding_mode` rather lengthy

`p+TAB+ENTER`

**Comment by OverLordGoldDragon** [comment]
> `torch.Conv1D` has `padding_mode`

[librosa](https://librosa.org/doc/latest/generated/librosa.stft.html)

**Comment by OverLordGoldDragon** [comment]
Numpy and TensorFlow use `mode`, TensorFlow's `Conv1D` uses `padding`, librosa uses `pad_mode` - can't match them all. I find `pad_mode` more attractive per brevity, changed to it for now - but I'm fine if we insist on anything else.

> unit tests

Unless you had something in mind, I don't see a need here; "testing" wasn't the best summary of my bullet, rather someone familiar with `1d` should see if I missed anything (but I can do later). With JTFS, both coverage and correctness are accounted since JTFS defaults to `zero`.

**Comment by lostanlen** [comment]
ok nevermind let's do `pad_mode`

I'm under the impression that this branch is not covered by unit tests?

```python
# check pad_mode
if self.pad_mode not in ('reflect', 'symmetric', 'zero'):
    raise ValueError("`pad_mode` must be one of: reflect, symmetric, "
                     "zero (got %s)" % self.pad_mode)
```

**Comment by OverLordGoldDragon** [comment]
@lostanlen You're right.

For such things I'd recommend a dedicated `test_exceptions.py`; Kymatio's coverage currently sits at 87%, these arg checks do indeed add up.

**Comment by lostanlen** [comment]
https://github.com/OverLordGoldDragon/kymatio/pull/4

yes a dedicated test file for exceptions would be nice
but for now i have put it under `test_numpy_scattering1d.py`

**Comment by janden** [comment]
Are we sure that `zero`, `reflect` and `symmetric` all are valid padding modes for the various frontends? It seems to line up for TF and NP, but Torch doesn't seem to have `zero` and `symmetric`. (Also I find `reflect` and `symmetric` to be synonymous, but perhaps that's just me…) We should maybe also specify in the docs what these different padding modes mean.

**Comment by OverLordGoldDragon** [comment]
> `symmetric`

@janden Just noticed torch lacks this... but `zero` is supported through `constant` (handled internally). Will change; currently only needed addon is `zero` anyway.

As for docs, where do I put them? Arg doc layout is strange.

**Comment by janden** [comment]
> @janden Just noticed torch lacks this... but `zero` is supported through `constant` (handled internally). Will change; currently only needed addon is `zero` anyway.

Makes sense. What about `symmetric`?

> As for docs, where do I put them? Arg doc layout is strange.

Try here: https://github.com/kymatio/kymatio/blob/776ad5988728117dba73a9a8c7a6cbef0f88593a/kymatio/scattering1d/frontend/base_frontend.py#L279

**Comment by OverLordGoldDragon** [comment]
@janden Only `zero` and `reflect` will be supported. Alts:

 1. Check per-backend
 2. Allow `lambda`

Idea with `2` is, instead of passing `scattering1d(pad=self.backend.pad)`, we pass `pad=self.pad_fn`, where

```python
self.pad_fn = (pad_mode if isinstance(pad_mode, FunctionType) else
               self.backend.pad)
```
and we document `pad_mode` func to have input signature `pad_fn(x, pad_left, pad_right, pad_mode)`.

Simple to implement, maximum flexibility. Should I proceed?

**Comment by lostanlen** [comment]
i like it

**Comment by janden** [comment]
> Only `zero` and `reflect` will be supported.

Ok got it.

> Allow `lambda`

I'm not sure. If you think it could be done cleanly, we can try.

The issue would be that the scattering1d core would try to pass `pad_mode` as an argument to that lambda function. It might be simpler to just do a partial substitution in the frontend and pass the padding function (with fixed padding mode) to the core code without `pad_mode` as another argument. Does that make sense?

**Comment by OverLordGoldDragon** [comment]
The workaround is nice; if looks good I'll do docs/frontends @janden

**Comment by OverLordGoldDragon** [comment]
And what of _minimum_ `T`? If it's any `T > 0`, subsampling logic needs to account for the case of `supp(phi) > supp(psi[-1])`.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 1]
```diff
@@ -46,18 +46,22 @@ def pad(cls, x, pad_left, pad_right):
...
+            name of padding to use.
         Returns
         -------
         output : tensor
             The tensor passed along the third dimension.
         """
         if (pad_left >= x.shape[-1]) or (pad_right >= x.shape[-1]):
             raise ValueError('Indefinite padding size (larger than tensor).')
+        if pad_mode == 'zero':
+            pad_mode = 'constant'
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/backend/numpy_backend.py:59]: Why dont we just name pad_mode to constant from the start?

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 2]
```diff
@@ -51,6 +54,21 @@ def build(self):
         else:
             raise ValueError("shape must be an integer or a 1-tuple")
 
+        # check `pad_mode`, set `pad_fn`
+        if isinstance(self.pad_mode, FunctionType):
+            def pad_fn(x):
+                return self.pad_mode(x, self.pad_left, self.pad_right)
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/base_frontend.py:60]: Why?

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 3]
```diff
@@ -175,6 +193,13 @@ def output_size(self, detail=False):
             `'array'`, the output is a large array containing the
             concatenation of all scattering coefficients. Defaults to
             `'array'`.
+        pad_mode : str (default 'reflect') / function, optional
+            Name of padding scheme to use, one of (`x = [1, 2, 3]`):
+                - zero:    [0, 0, 1, 2, 3, 0, 0, 0]
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/base_frontend.py:198]: just name it constant?

**Review by MuawizChaudhary — COMMENTED** [review]
Why dont we rename `padding_mode='zero'` to `'constant'`?

**Comment by MuawizChaudhary** [comment]
> > Only `zero` and `reflect` will be supported.
> 
> Ok got it.
> 
> > Allow `lambda`
> 
> I'm not sure. If you think it could be done cleanly, we can try.
> 
> The issue would be that the scattering1d core would try to pass `pad_mode` as an argument to that lambda function. It might be simpler to just do a partial substitution in the frontend and pass the padding function (with fixed padding mode) to the core code without `pad_mode` as another argument. Does that make sense?

not a fan of this lambda stuff.

**Comment by MuawizChaudhary** [comment]
We'll refactor the padding in a future release.
